[FEAT] New LoRA Initialization Method: Explained Variance Adaptation

sirluk opened the pull request:
### Description

In our work "One Initialization to Rule them All: Fine-tuning via Explained Variance Adaptation" ([paper](https://arxiv.org/abs/2410.07170)) we introduce a new data-driven initialization method for LoRA adapters. We will present this work at [ENLSP](https://neurips2024-enlsp.github.io/#)  (oral) and [AFM](https://adaptive-foundation-models.org/call-for-papers.html) at NeurIPS'24.

Eva initializes LoRA in a data-driven manner based on information of the downstream data using SVD. It also adaptively allocates ranks throughout the model.

### Implementation Details

1. As the initialization is data-driven we created a new config class `EvaConfig` which at minimum needs to receive a value for the `dataloader` argument. Other parameters have default values which worked well in our experiments. The instantiated EvaConfig is passed as a parameter to LoraConfig. (we took inspiration from the LoftQConfig implementation).
In LoraConfig "init_lora_weights" needs to be set to "eva"
2. We modify the `__init__` method of `LoraModel` to create a state dict for eva initialzation before initializing the parent class to be able to populate the `rank_pattern`argument in `LoraConfig`. This way we directly initialize with a modified rank distribution allowing us to directly load the eva_state_dict.
3. All other code necessary for the initialzation is self-contained in peft/tuners/lora/eva.py
4. Under examples/eva_finetuning/eva_finetuning.py we add an example script for eva initialization

### Open Questions

- Not sure if `self.model.load_state_dict(eva_state_dict, strict=False)` is always a valid way to initialize. E.g. when the model is wrapped with torch.compile
- We currently only tested a single GPU setting. We are not sure what the best way is to have multi-gpu support (only run init on rank 0 and copy to other ranks when done?)

### Example Usage

```python
eva_config = EvaConfig(
    dataloader = dataloader
)
peft_config = LoraConfig(
    r = 16,
    lora_alpha = 1,
    target_modules = ['q_proj', 'k_proj', 'v_proj', 'o_proj'],
    init_lora_weights = "eva",
    eva_config = eva_config,
)
peft_model = get_peft_model(model, peft_config)
```

BenjaminBossan commented:
Thanks a lot for this PR. I only skimmed it for now and have yet to read the details from the paper but this looks quite promising!

When looking at the implementation, I saw that the way the Eva weights are initialized sits a bit "outside" of how we normally do this in PEFT. The normal way is to have this on the layer level, check for instance here:

https://github.com/huggingface/peft/blob/5758a7eb1c156b2c513930af666acb2d4a2cb3c3/src/peft/tuners/lora/layer.py#L128-L139

My guess why this is implemented differently here is that there is currently one pass through the modules that determines the individual ranks and creates the state dict. If this was refactored such that the weights were initialized on the layer level as usual, we would need two passes, one to adjust the ranks in the config and then the actual weight initialization on the layer level. Would it be possible to refactor the code to enable that?

My main concern with the current approach is that the weight initialization currently is happening in one place of the code but now there would be another place, making the code more disorganized. Second, IIUC, we would initialize the weights once and then override them with Eva parameters, which is wasteful. Finally, if we do this on the layer level, that could solve the multi-GPU issue, right?

Apart from this, what I noticed while skimming is that there is a big class `IncrementalPCA` which looks like a rewrite of the equivalent [sklearn class](https://scikit-learn.org/1.5/modules/generated/sklearn.decomposition.IncrementalPCA.html) using torch. My question is: Did you implement this from scratch or is it taken from somewhere? This has implications on how well this is tested etc.

I also had difficulties understanding the `SVDHook` and `HashHook`, could you please add a small docstring? Are these hooks permanently added?

Regarding `torch.compile`, we could add a test for that and see if it works. But this is not a necessity for this PR to land in PEFT, just something that's nice to have.

Going forward, we'll also have to add a few unit tests, but I'm fine with iterating on the implementation first and then adding the tests.

sirluk commented:
Thanks a lot for the quick feedback @BenjaminBossan !

I added docstrings to `SVDHook`and `HashHook`, hope this makes their purpose a bit more clear.

Regarding your other concerns about the implementation:

- I agree its not that our proposed init method is not on a layer level as the other inits. I thought a lot about where in the code to put the init and unfortunately I think it would be quite difficult to put the init directly in the layer. (or at least I haven't found an elegant way to do it). The way eva is initialized is that we do multiple forward passes through the model, and incrementally perform SVD on the input activations of the target layers (usually more than 10 forward passes). The resulting singular vectors are used to initialize the Lora A matrices. So unfortunately 2 passes would also not be enough. We would also need to somehow pass an instance of `EvaConfig` to the `LoraLayer`.
I did however add a statement in the `update_layer` method to intialize the Lora B matrices with zeros as needed for eva. For this reason we do not initialize weights twice as the A matrices are not initialized in the `update_layer` method.

- Regarding the new eva code. IncrementalPCA is indeed largely take from the [sklearn class](https://github.com/scikit-learn/scikit-learn/blob/d5082d32d/sklearn/decomposition/_incremental_pca.py#L19). So a lot of the code is directly copied from there. What we added is seamless gpu support by setting the device for all internal variables / buffers to the device of the first input tensor. We also integrated support for torch.svd_lowrank to significantly speed up computations. We did several tests internally if the results are equivalent when running this on gpu compared to running the sklearn implementation on cpu.

sirluk commented:
### Implementation

I was wondering already if having the dataloader in the confige is a good idea since you probably only want primitive types there. I think your suggested solution is a sensible one. I gave it some thought and dont really have a better suggestion following the restrictions we discussed.

A few questions for clarification:
- Would eva still be a valid value for `init_lora_weights` in `LoraConfig`. I am a bit worried about discoverability of eva if `LoraConfig` would then just be normal lora settings and someone needs to know about the function `initialize_lora_eva_weights` to use eva.
- In your suggested solution, would sample data not be a dataloader?
- Is it necessary to define a custom forward function for the eva init? Usually it should be anyway the same inputs that the model receives during normal finetuning. I did adapt the line you mentioned filtering out the label key from the inputs dict.
- One question about how we get the layer activations for SVD. In the forward hook `SVDHook` we currently have it implemented like this
```python
try:
    states = input.detach()
except AttributeError:
    states = input[0].detach()
```
So for cases where we input is a tuple, we just take the first value. Not super important at this stage I guess but I just wanted to ask if you think we should add an argument so the user can define what the inputs we should use are.

### IncrementalPCA

It's a good idea to take some from sklearn. All tests for this class are in this [file](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/decomposition/tests/test_incremental_pca.py). Right now there is nothing like this available anywhere. I built my own package [torch-incremental-pca](https://github.com/sirluk/pytorch_incremental_pca) but I just assumed you dont want any dependencies to small packages like this so I just copied the class over here. I did however create a pull request to have this feature implemented in pytorch directly so in the future we might be able to remove this from peft again.

sirluk commented:
@BenjaminBossan before going ahead with this implementation of calling a function/method after `get_peft_model` is called I just wanted to propose one alternative idea to get your thoughts on it.

Instead of using `LoraModel` we create a subclass `EvaModel`. This model class initially would not have any adapter layers but we would add the forward hooks needed for the SVD computations. In the `__init__` we would have something like

```python
class EvaModel(LoraModel):

	def __init__(self, model, config, adapter_name)
		self.model = model
		for layer in config.target_modules:
			module = get_module(self.model, layer)
			module.register_forward_hooks(SVDHook(config))
```

During the first few forward passes of finetuning we would not actually update any weights but use the activations to compute the eva state dict. Once SVD is converged for all layers we would switch the internal state of the model to finetuning (with some boolean variable) and at that point only add the adapters with the appropriate eva values for the lora A matrices.

Hope this makes sense. That way we could leverage the finetuning loop directly. It seem a bit more seamless and works without needing any additional function calls after `get_peft_model`. But of course there might be some hidden issues I am not considering at the moment.

BenjaminBossan commented:
I like the proposal to seamlessly integrate this into the training loop. I'm not sure if we need a separate class for that, but you could come up with a prototype and I can check if there is a good way to integrate it without a separate class.

With this, we would need to make sure that it works with vanilla training loops but also frameworks like `Trainer` from transformers. If it does, I think this could be the optimal solution.

sirluk commented:
@BenjaminBossan I started implementing this approach of using the first few batches for the initialization and unfortunately I realised that there are at least 2 major issues with it:
- The LoRA layers must be initialized at the beginning of training so that they can be passed to the optimizer. This is especially important if it needs to work out of the box with the HuggingFace Trainer. However, this means that the LoRA weights would take up unnecessary GPU memory that would be needed for the SVD computation. This is btw also an issue with the solution you proposed of doing a normal LoRA init first and then calling a function to initalize eva.
- Further, if you use a learning rate schedule, it would already adjust the learning rate while you are still executing the SVD steps. This means that when you start fine-tuning, the learning rate is different from the intended initial learning rate. Especially problematic if there is a warmup I would assume.

These issues makes me wonder if some variant if my initially proposed solution would be feasible (computing the eva state dict before initializing the adapters)

One idea I would have which avoids passing a dataloader or something similar through the config would be to enable the `get_peft_model`function to accept some optional keyword arguments through `**kwargs` and passing the dataloader as well as a model `forward_fn` this way.

```python
def get_peft_model(
	model: PreTrainedModel,
	peft_config: PeftConfig,
	adapter_name: str = "default",
	mixed: bool = False,
	autocast_adapter_dtype: bool = True,
	revision: Optional[str] = None,
	**model_init_kwargs
):
   ...
   PeftModel(model, peft_config, adapter_name=adapter_name, autocast_adapter_dtype=autocast_adapter_dtype, **model_init_kwargs)
   ...



peft_model = get_peft_model(model, peft_config, eva_dataloader=dataloader, eva_forward_fn=eva_forward_fn)
```

We also don't directly have to modify the `__init__` method of `LoraModel` but define a `_pre_injection_hook` for `LoraModel` which should also do the trick.

Let me know what you think about the issues I mentioned with the previously proposed approaches.

sirluk commented:
@BenjaminBossan thanks for your suggestions, I implemented a prototype solution now for the discussed approach. A few questions:
- I moved the IncrementalPCA class which has been adapted from sklearn to peft.utils.incremental_pca.py because I thought it is not strictly part of eva and in theory could be used for other things, does that make sense?
- Ideally users should be able to do `from peft import `initialize_lora_eva_weights`. Would it be suitable to then put this function in peft.utils or should it be in peft.tuners.lora. I didnt see any other functions being pulled directly into the peft namespace from the tuners directory.
- I was wondering if we should use `low_cpu_mem_usage=True` as a default for eva. That way we would even need to have it as an argument for `get_peft_model`

BenjaminBossan commented on src/peft/tuners/lora/config.py:
```diff
@@ -357,6 +396,10 @@ def __post_init__(self):
+            
+        elif self.init_lora_weights == "eva":
+            if self.eva_config is None:
+                raise ValueError("`eva_config` must be specified when `init_lora_weights` is 'eva'.")
```
Let's also check the reverse, i.e. if `self.init_lora_weights != "eva"`, there should be no `eva_config`.

BenjaminBossan reviewed (commented):
> I implemented a prototype solution now for the discussed approach.

Thanks a lot. I think the API is decent and given the trade-offs we discussed, probably a good compromise.

> * I moved the IncrementalPCA class which has been adapted from sklearn to peft.utils.incremental_pca.py because I thought it is not strictly part of eva and in theory could be used for other things, does that make sense?

Yes, that's fine. In the end, I think it doesn't really matter. What's important is that it is not in the `peft` namespace (i.e. `from peft import IncrementalPCA` should *not* be possible), apart from that it doesn't matter too much. We can always move it later.

> * Ideally users should be able to do `from peft import `initialize_lora_eva_weights`. Would it be suitable to then put this function in peft.utils or should it be in peft.tuners.lora. I didnt see any other functions being pulled directly into the peft namespace from the tuners directory.

We do pull objects from `peft.tuners` into the `peft` namespace, otherwise `from peft import LoraConfig` would not be possible :) As for `initialize_lora_eva_weights`: Right now, this only works for LoRA and it's not planned to change that, right? Therefore, I would keep it in the `tuners/lora` directory. There, import it into `__init__`, then from there into the `__init__` from `tuners/` and from there into the `__init__` from `peft`.

> * I was wondering if we should use `low_cpu_mem_usage=True` as a default for eva. That way we would even need to have it as an argument for `get_peft_model`

How exactly would you do that, given that the `get_peft_model` is done by the user?

sirluk commented:
@BenjaminBossan thanks again for your feedback!

> We do pull objects from `peft.tuners` into the `peft` namespace, otherwise `from peft import LoraConfig` would not be possible :)

I pulled `initialize_lora_eva_weights` and `get_eva_state_dict` into the peft namespace now. I asked because I only saw classes being pulled from tuners, no functions so I was wondering if there is some rule for that.

> How exactly would you do that, given that the `get_peft_model` is done by the user?

I thought we could just not have a check in the `LoraModel.__init__`
```python
low_cpu_mem_usage = init_lora_weights="eva" or low_cpu_mem_usage
```

Regarding tests I adapted some tests from sklearn for the incremental_pca class in `tests/test_incremental_pca.py`. Could you maybe advice me what the minimum test setup should look like for eva.

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,463 @@
+    if device is not None:
+        recursive_apply(model, device)
+    
+    model.disable_adapter()
```
This can be removed, right?

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,463 @@
+    model,
+    peft_config,
+    dataloader,
+    forward_fn: callable = lambda model, inputs: model(**inputs),
```
Let's make this optional, and if `None` is passed, define a local function and use that for `forward_fn`. Let's avoid assigning a `lambda` function.

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,463 @@
+    adapter_name: (str)
+        The name of the adapter to initialize the weights for.
+    prepare_activations_fn: (Union[callable, Dict[str, callable], None])
+        If a layer recieves multiple inputs as a list, per default the first input is used.
```
```suggestion
        If a layer receives multiple inputs as a list, per default the first input is used.
```

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,463 @@
+import torch
```
Let's add the license notice.

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,463 @@
+    layer_hook_map = {**dict(zip(hooks.keys(), hooks.keys())), **equal_inputs_map}
+
+    # start svd calculation
+    pbar = tqdm(iter(cycle(dataloader)), position=0, leave=False)
```
Just wondering: Should we make the progress bar optional, e.g. via an extra argument? Maybe some users don't want it.

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,463 @@
+
+        for name, hook in hooks.items():
+            module = model.get_submodule(name)
+            module._forward_hooks.clear()
```
Hmm, should we call `hook.remove()` instead? What if there are other, unrelated forward hooks on the model?

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,463 @@
+    device = get_device_with_meta_params(model)
+    model.eval()
+
+    hooks = {}
```
Do we need a check at the end that all the hooks have been cleaned up or is that not necessary?

sirluk replied:
good idea! I implemented a check now
```python
    # check all svd hooks have been removed
    remaining_svd_hooks = {n for n, m in model.named_modules() for v in m._forward_hooks.values() if isinstance(v, SVDHook)}
    if len(remaining_svd_hooks) > 0:
        raise ValueError(f"All SVD hooks should have been removed but the following layers still have active hooks: {remaining_svd_hooks}")
```

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,463 @@
+    """
+    Get the device of the model's parameters. Useful if some parameters are on meta device.
+    """
+    devices = list(set([p.device for p in model.parameters() if str(p.device) != "meta"]))
```
You can use a set comprehension: `{p.device for ...}`.

BenjaminBossan commented on src/peft/tuners/lora/config.py:
```diff
@@ -69,6 +69,31 @@ class LoftQConfig:
+    device: str = field(default="cuda", metadata={"help": "Device to use for EVA initialization"})
+
+    def __post_init__(self):
+        assert self.rho >= 1.0, "early_stop_rho must be >= 1"
```
Let's completely avoid all `assert`s except for tests. Instead, raise a `ValueError` here. Same for the other asserts used below.

BenjaminBossan commented on src/peft/tuners/lora/config.py:
```diff
@@ -69,6 +69,31 @@ class LoftQConfig:
+@dataclass
+class EvaConfig:
+    """
+    This is the sub-configuration class to store the configuration for a data-drive initialization via EVA.
```
Let's add a link to the paper here as well. Also, what's always nice to have for users is to give some indication when/how to change the parameters. E.g. "increase rho if ...".

BenjaminBossan commented on src/peft/tuners/lora/config.py:
```diff
@@ -69,6 +69,31 @@ class LoftQConfig:
+    use_label_mask: bool = field(default=False, metadata={"help": "Use label mask for EVA initialization"})
+    label_mask_value: int = field(default=-100, metadata={"help": "if use_label_mask=True the value to look for to mask out ignored tokens"})
+    whiten: bool = field(default=False, metadata={"help": "Apply whitening to singular vectors"})
+    device: str = field(default="cuda", metadata={"help": "Device to use for EVA initialization"})
```
Do we need this? I don't think so.

BenjaminBossan reviewed (changes requested):
Thanks a lot for these updates to the PR.

Please always run `make style` after you made changes to the code so that the linter will pass.

> I pulled `initialize_lora_eva_weights` and `get_eva_state_dict` into the peft namespace now. I asked because I only saw classes being pulled from tuners, no functions so I was wondering if there is some rule for that.

Ah okay, no worries, there are no rules about that.

> I thought we could just not have a check in the `LoraModel.__init__`
> 
> ```python
> low_cpu_mem_usage = init_lora_weights="eva" or low_cpu_mem_usage
> ```

I see. I think I'd rather not automatically switch this, maybe users have a specific reason why they need `low_cpu_mem_usage=False` despite using Eva. What you could do is give a warning if you find this, advising users that passing `low_cpu_mem_usage=True` is most likely what they want.

> Regarding tests I adapted some tests from sklearn for the incremental_pca class in `tests/test_incremental_pca.py`.

I think what you have is more than enough, the test coverage for the class is very good. Let's just add a comment on where you adopted these tests from.

> Could you maybe advice me what the minimum test setup should look like for eva.

This is a very good question. From my understanding, we can't really check the adjusted weights to say that after applying Eva, the weights should have this or that property. Just as an example, for PiSSA/OLoRA, we have tests that they reduce the quantization error. But I don't think we can have something similar here, right?

What we could do is just calling `initialize_lora_eva_weights` with a couple of different options to ensure that there is no error being raised. This helps with the test coverage but it wouldn't really help with ensuring that Eva actually works as described in the paper.

Maybe we can expand the included example to replicate one of the findings from the paper. That way, we would have a high confidence that it works as promised, WDYT about that?

Alternatively, we could create small self-contained training runs in the test, using a small model and dataset. Then we would train the model once with and once without Eva. The test would be e.g. that with Eva, the model converges faster. This would be a more functional test but of course it has the danger of being flaky/seed dependent and could be slow to run.

sirluk commented on src/peft/tuners/lora/config.py:
```diff
@@ -359,6 +398,11 @@ def __post_init__(self):
+        elif self.init_lora_weights == "eva" and self.eva_config is None:
+            raise ValueError("`eva_config` must be specified when `init_lora_weights` is 'eva'.")
+        elif self.init_lora_weights != "eva" and self.eva_config is not None:
+            warnings.warn("`eva_config` specified but will be ignored when `init_lora_weights` is not 'eva'.")
```
I thought maybe when someone is comparing multiple lora init methods they can just have the eva_config always inside the lora_config and need to just change the init_lora_weights parameter

BenjaminBossan replied:
I see, I guess that could happen. Then I'm fine with leaving it as is.

sirluk commented:
@BenjaminBossan thanks for all your suggestions, I integrated all of it into the pull request.

Regarding testing I added a new class `TestEvaInitialization` into `tests_initialization.py`. The tests basically run eva initialization three times for different seeds and check that the average cosine similarity between the same svd component from different seeds is above a certain threshold. I defined two three test cases. I fear doing a finetuning run or replicating one of the findings of the paper would probably take up too much resources / take to long for a testcase.

1. Check the `get_eva_state_dict` function which returns the state_dict for eva. We use a non-peft model because testing peft models is covered with the next test
2. Check `initialize_lora_eva_weights` with a LoRA model which internally also calls `get_eva_state_dict`
3. Check the `__post_init__` method of `EvaConfig` for invalid settings.

During testing I noticed a potential issue with how the `loftq_config` field is defined in `LoraConfig`. I added an [issue](https://github.com/huggingface/peft/issues/2170) and could also create a separate pull request for it.

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,505 @@
+        if not isinstance(module, LoraLayer):
+            continue
+        if name in eva_state_dict:
+            w = eva_state_dict.pop(name)
```
Let's use named parameters here instead of relying on position. I think this can also simplify the code: Outside of the loop, create a `kwargs` dict with all the arguments that are the same in both calls, then only pass the arguments that differ.

sirluk replied:
make sense, done

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,505 @@
+            indices = get_indices_fn(inputs, peft_config)
+
+        for name in list(hooks.keys()):
+            hook, handle = hooks.get(name)
```
Why not `hooks[name]`?

sirluk replied:
switched to `hooks[name]`

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,505 @@
+    }
+    if len(remaining_svd_hooks) > 0:
+        raise ValueError(
+            f"All SVD hooks should have been removed but the following layers still have active hooks: {remaining_svd_hooks}"
```
What should a user do in this case? IIUC, this would be a bug and they should report an issue on GH. If that's the case, please add that instruction to the error message.

sirluk replied:
added a message that an issue should be created

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,505 @@
+        if rank == 0:
+            continue
+        hook = hooks[layer_hook_map[name]][0]
+        assert torch.all(hook.converged[:rank])  # this should never happen because we check for convergence
```
Let's raise a proper error here.

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,505 @@
+    PCA method to compute the SVD components. The function also checks for convergence of the computed components using
+    cosine similarity. The rank distribution for each layer is determined based on the explained variance ratio.
+
+    Parameters:
```
You're using the numpydoc docstring convention here (which sklearn also uses), but we use the (google?) docstring convention in PEFT. Could you please adjust it here and throughout?

sirluk replied:
done

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
````diff
@@ -0,0 +1,505 @@
+        ```
+
+    Returns:
+    -------- eva_state_dict: (dict)
````
missing line break

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,505 @@
+    layer during the forward pass of a neural network. The hook also tracks convergence of the computed components
+    using cosine similarity between the current and previous components.
+
+    Attributes:
```
"Attributes" should be "Parameters" but as mentioned elsewhere, we use a different docstring convention altogether.

BenjaminBossan commented on src/peft/tuners/lora/config.py:
```diff
@@ -69,6 +69,37 @@ class LoftQConfig:
+
+    def __post_init__(self):
+        if self.rho < 1.0:
+            raise ValueError("`rho` must be >= 1.")
```
Would it make sense to also add a check on tau?

sirluk replied:
yes, for sure that makes sense, I added a check

BenjaminBossan commented on tests/test_initialization.py:
```diff
@@ -1544,3 +1550,112 @@ def new_state_dict():
+
+    @pytest.fixture
+    def dataset(self, tokenizer):
+        dataset = load_dataset("Rowan/hellaswag", split="train")
```
I'd rather not use this dataset here in the tests, as it's downloaded from an external source. Would the test also work with, for instance, `"ybelkada/english_quotes_copy"`, which already use in many other tests? Maybe that can also help speed up testing.

sirluk replied:
done

BenjaminBossan commented on tests/test_initialization.py:
```diff
@@ -1544,3 +1550,112 @@ def new_state_dict():
+    @pytest.fixture
+    def model(self):
+        model = AutoModelForCausalLM.from_pretrained("openai-community/gpt2")
+        model.transformer.h = torch.nn.ModuleList([model.transformer.h[0]])  # truncate to 1 layer
```
Just wondering if we should maybe keep 2 layers to cover edge cases a little bit better while still being fast.

sirluk replied:
thats a good point!

BenjaminBossan commented on tests/test_initialization.py:
```diff
@@ -1544,3 +1550,112 @@ def new_state_dict():
+    @pytest.fixture
+    def peft_config(self):
+        return LoraConfig(
+            r=16, lora_alpha=1, target_modules=["c_attn"], init_lora_weights="eva", eva_config=EvaConfig(rho=2)
```
Is there a specific reason for these a bit strange parameters: r=16, alpha=1?

sirluk replied:
its the setting for most experiments in our paper

BenjaminBossan commented on tests/test_initialization.py:
```diff
@@ -1544,3 +1550,112 @@ def new_state_dict():
+    def peft_model(self, model, peft_config):
+        return get_peft_model(deepcopy(model), peft_config, low_cpu_mem_usage=True)
+
+    def test_eva_state_dict_consistency(self, model, dataset, peft_config):
```
Could you please add a short explanation of what is tested here and why?

BenjaminBossan commented on tests/test_initialization.py:
```diff
@@ -1544,3 +1550,112 @@ def new_state_dict():
+            f"is not greater than {self.COSINE_SIMILARITY_THRESHOLD}"
+        )
+
+    def test_eva_initialization_consistency(self, peft_model, dataset, peft_config):
```
Could you please add a short explanation of what is tested here and why?

BenjaminBossan commented on tests/test_initialization.py:
```diff
@@ -1544,3 +1550,112 @@ def new_state_dict():
+    COSINE_SIMILARITY_THRESHOLD = 0.75
+    NUM_SEEDS = 3
+    BATCH_SIZE = 4
+    MAX_LENGTH = 512
```
Maybe we can go down here for speed?

BenjaminBossan commented on examples/eva_finetuning/eva_finetuning.py:
```diff
@@ -0,0 +1,68 @@
+
+initialize_lora_eva_weights(peft_model, peft_config, dataloader, device=svd_device)
+
+# from this point on, you can use the model as you would use a normal LoRA model
```
Just wondering: Would it make sense to actually include the fine-tuning code here, ideally using an example from/close to the paper. Or alternatively, add a second script with a complete example?

sirluk replied:
yes I guess it makes sense to have a complete example, anyway its basically just initializing the Trainer class. I added the necessary code to run the full example.

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,505 @@
+            )
+            module.lora_A[adapter_name].weight.copy_(w)
+        else:
+            module.update_layer(
```
What would be a situation when this can occur? During the unit testing, it is not covered.

sirluk replied:
Its mostly because we dont support embedding and conv layers at the moment. So if adapters are added to those layers we cant initialize it with layer so the next best thing would be to do normal lora initalization by default.

BenjaminBossan commented on tests/test_initialization.py:
```diff
@@ -1544,3 +1550,112 @@ def new_state_dict():
+        with pytest.warns(
+            UserWarning, match="`eva_config` specified but will be ignored when `init_lora_weights` is not 'eva'."
+        ):
+            LoraConfig(init_lora_weights=True, eva_config=EvaConfig())
```
Can we have small tests for:

- `whiten=True`
- `prepare_activations_fn` being a mapping

sirluk replied:
I added a whitening setting for `test_eva_state_dict_consistency`. To test `prepare_activations_fn` I added a test `test_eva_state_dict_prepare_activations_mapping`

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,505 @@
+    inputs = next(iter(dataloader))
+    forward_fn(model, move_inputs_to_device(inputs, device))
+    hash_dict = {k: h[0].hashed_inputs[0] for k, h in hooks.items()}
+    equal_inputs_map = {vv: v[0] for v in find_equal_values(hash_dict).values() for vv in v[1:]}
```
I don't really understand what should happen to have `equal_inputs_map`, could you please add a comment?

sirluk replied:
Added a comment. It contains information about which layers recieve the same input activations (e.g. queries, keys and values). For those groups we only need to compute the svd once.

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,505 @@
+
+        if previous_components is not None:
+            components = self.svd.components_
+            if len(components.shape) == 1:
```
Could you add a short comment when this would happen?

sirluk replied:
done, its only the case during the first svd step

BenjaminBossan reviewed (changes requested):
Thanks a lot for all the updates. I did a more in-depth review now and found a few potential issues, please check my comments. Apart from those, could you please:

1. Always run `make style` after all changes are done
2. Resolve the merge conflict.
3. Clarify: Should it work with quantization like bnb? I think so but maybe it can be made explicit or even a small test be added.

sirluk commented on examples/eva_finetuning/README.md:
```diff
@@ -0,0 +1,101 @@
+# EVA: Efficient Vector Adaptation
+## Introduction ([Paper](https://arxiv.org/abs/2404.02948), [code](https://github.com/GraphPKU/EVA))
```
thanks, fixed the link! I would actually leave the code link in, we have alot of scripts there which allows people to reproduce the results from the paper.

sirluk commented on examples/eva_finetuning/README.md:
```diff
@@ -0,0 +1,101 @@
+When `initialize_lora_eva_weights` is called, it will load the components calculated by EVA into the model. After this continue with standard LoRA finetuning.
+
+## Getting eva_state_dict without loading the adapter weights
+If you want to get the eva_state_dict without loading the adapter weights, you can do the following:
```
done

sirluk commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,505 @@
+    else:
+
+        def _check_fn(name, module):
+            return any(name.endswith(t) for t in peft_config.target_modules)
```
thanks for the suggestion, I improved this part a bit now.
- for peft models I now just check `hasattr(module, "base_layer")` and I defined a new variable `UNSUPPORTED_LORA_MODULES` which are not used for eva (atm lora.Embedding and lora.Conv layers are not supported)
- for non-peft models I am using `check_target_module_exists` now and filter for only `torch.nn.Linear` and `transformers.pytorch_utils.Conv1D` layers

sirluk commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,505 @@
+    """
+    orig_device = get_device_with_meta_params(model)
+    if device is not None:
+        recursive_apply(model, device)
```
I guess its fine to remove it and rely on the user to move the model to the appropriate device before calling this function

sirluk commented on src/peft/tuners/lora/eva.py:
````diff
@@ -0,0 +1,505 @@
+        ```
+
+    Returns:
+    -------- model: (torch.nn.Module)
````
I guess `make style` did this as I was using the wrong docstring format

sirluk commented:
@BenjaminBossan I integrated all of your suggestions in the code, many thanks (for details I left comments to the separate conversations you opened about suggested changes).

Two general questions:
- I included a gpu-dependent test in test_initialization with `pytest.mark.skipif(not torch.cuda.is_available(), reason="test requires a GPU")` to test bitsandbytes. Is that fine to have all EVA tests in one class in this script or would I need to move this test to `test_common_gpu` or `test_gpu_examples`?
- under `examples/eva_finetuning/README.md` I added sections using the [[autotest]] command to automatically use the docstrings of the eva related functions. Is that fine or does that not work for the examples section?

BenjaminBossan commented on src/peft/utils/incremental_pca.py:
```diff
@@ -0,0 +1,321 @@
+
+        if lowrank:
+            if lowrank_q is None:
+                assert (
+                    n_components is not None
+                ), "n_components must be specified when using lowrank mode with lowrank_q=None."
+                lowrank_q = n_components * 2
+            assert lowrank_q >= n_components, "lowrank_q must be greater than or equal to n_components."
```
Let's raise proper errors instead of using asserts.

BenjaminBossan commented on src/peft/utils/incremental_pca.py:
```diff
@@ -0,0 +1,321 @@
+            return last_mean, last_variance, last_sample_count
+
+        if last_sample_count > 0:
+            assert last_mean is not None, "last_mean should not be None if last_sample_count > 0."
+            assert last_variance is not None, "last_variance should not be None if last_sample_count > 0."
```
Let's raise proper errors instead of using asserts.

BenjaminBossan reviewed (changes requested):
Thanks a lot for your excellent work. Not much is missing to be able to merge this.

> Eva works out of the box with bitsandbytes. I included a gpu-dependent test in test_initialization with pytest.mark.skipif(not torch.cuda.is_available(), reason="test requires a GPU") to test bitsandbytes. Is that fine to have all EVA tests in one class in this script or would I need to move this test to test_common_gpu or test_gpu_examples?

For this test to be actually run on our nightly CI with GPUs, you'd have to move it, I think `test_gpu_examples.py` is the right place (just create a new test class there).

> under examples/eva_finetuning/README.md I added sections using the [[autotest]] command to automatically use the docstrings of the eva related functions. Is that fine or does that not work for the examples section?

Hmm, not sure what you mean, I'm not familiar with `autotest`. Could you please elaborate?

> To make using eva more seamless, do you think its a good idea to just have to specify init_lora_weights="eva" and if  EvaConfig is not defined / none it is initialized with the defaults (i.e. EvaConfig())?

I think it makes sense to do that if the default config works well enough out of the box. If we want the users to actively set the Eva parameters, it is better to make this step explicit. You can judge better than I can whether the default parameters should work well enough.

sirluk commented on src/peft/mapping.py:
```diff
@@ -162,6 +163,8 @@ def get_peft_model(
             The revision of the base model. If this isn't set, the saved peft model will load the `main` revision for
             the base model
+        low_cpu_mem_usage (`bool`, `optional`, defaults to `False`):
+            Create empty adapter weights on meta device. Useful to speed up the loading process.
```
added this

sirluk commented on examples/eva_finetuning/eva_finetuning.py:
```diff
@@ -0,0 +1,87 @@
+# setup peft config
+eva_config = EvaConfig(rho=rho)
+peft_config = LoraConfig(
+    r=rank, lora_alpha=alpha, target_modules=target_modules, init_lora_weights=True, eva_config=eva_config
```
thanks for pointing out this error!

sirluk commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,545 @@
+        else:
+            module.update_layer(r=peft_config.r, init_lora_weights=True, **update_layer_kwargs)
+            missing_eva_inits.append(name)
+            print(name, type(module))
```
Add a test for this `test_missing_eva_inits`. Since EVA at the moment does not support embedding layers for example, we are doing standard lora init if such layers have adapters.

sirluk commented:
Thank you again for your suggestions @BenjaminBossan, I worked them all into the pull request.

- I moved the test for bits and bytes to `test_gpu_examples.py`. It was necessary to duplicated some code but I guess thats fine.
- I updated the documentation a bit. I also added a section for eva under [package reference](https://github.com/sirluk/peft/blob/main/docs/source/package_reference/lora.md), please let me know if its okay like this. I think it would be beneficial to find info about the EVA functions in the documentation.
- Since my last message in this pull request I also made changes to the entrypoint `get_indices_fn` for customizing EVA. I made it a bit more generic and easier for users to understand. I renamed the callable argument in `get_eva_state_dict` and `initialize_lora_eva_weights` to `prepare_model_inputs_fn` now. It can be used to modify the model inputs and have access to the result in the SVD hooks. I adjusted the tests accordingly.
- I discussed the topic about a default `EvaConfig` with my coauthors. Some parameters like `rho` or `user_label_mask` have a setting that should work well for most applications but nevertheless it can be important to adjust the default settings. We were wondering if we should initialize a default `EvaConfig` if its not specified but raise a warning explaining that it might be important to change the parameters in some cases. In either case I also updated the docstring of `EvaConfig` to mention this.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_2142). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

BenjaminBossan commented on examples/eva_finetuning/eva_finetuning.py:
Nice example. I tried it locally with a smaller model and train loss with Eva enabled is consistently better than without it.

sirluk replied:
great to hear!

BenjaminBossan reviewed (changes requested):
Thanks a lot for all the updates. The CI is a bit flaky at the moment, so don't worry about some failing test runs.

While reviewing the code, I found one major blocker that needs to be addressed before we can merge. This issue is that Eva redistributes the LoRA ranks, right? However, this is not reflected in the `LoraConfig`. Therefore, if a user saves the model and then tries to load it later, they'll get errors, e.g.:

> size mismatch for base_model.model.model.layers.0.self_attn.o_proj.lora_A.default.weight: copying a param with shape torch.Size([19, 2048]) from checkpoint, the shape in current model is torch.Size([16, 2048])

So unless I'm missing something, the solution for this would be to get the new ranks for all the LoRA weights and store them in a dict mapping the module name to the rank. This dict should be used to override the `rank_pattern` attribute of the `LoraConfig`. This should allow to successfully load the model. Could you please take a look at that and also extend the tests to cover this?

Regarding the alpha values, I'm not sure if those also need to be adjusted. E.g. if I have a LoRA rank of 16 and alpha of 32, the LoRA scaling would be 32/16. If the rank is changed to 19 but alpha stays at 32, the scaling would decrease to 32/19. Probably it's not a big deal, but I wanted to mention that.

> I also added a section for eva under [package reference](https://github.com/sirluk/peft/blob/main/docs/source/package_reference/lora.md), please let me know if its okay like this. I think it would be beneficial to find info about the EVA functions in the documentation.

Yes, that should be okay. Unfortunately, the doc preview appears to be broken at the moment, but if there is anything wrong with the rendering there, it'll be easy enough to fix later.

> We were wondering if we should initialize a default EvaConfig if its not specified but raise a warning explaining that it might be important to change the parameters in some cases.

From my point of view, the current solution is good.

sirluk commented:
@BenjaminBossan thanks for raising the issue with saving/loading eva models, didnt think about this at first. It should be fixed now. I am updating the `rank_pattern` in `init_lora_eva_weights`.

I also found another potential issue that was not accounted for so far. It is possible that after eva initialization some adapter layers could get assigned rank 0 . In those cases we would need to remove the adapter layer again. We would also need to update `target_modules` in the peft config. I also added test cases for this and the above issue. To test the `rank=0` case it was necessary to move some functionality out of `init_lora_eva_weights` into a new function `load_eva_state_dict`. Otherwise it would not have been possible to modify the eva_state_dict (i.e. assigning one layer rank 0) in the test script. `init_lora_eva_weights` now just combines the two lower level functions `get_eva_state_dict` and `load_eva_state_dict`

Regarding alpha values, it should be fine as is for now. For the experiments in our paper we also did not adjust alpha for different ranks after the redistribution. I saw that the AdaLoRA implementation also did not adjust alpha for different ranks, is that right? We might look into this in the future.

Regarding docs, I built it locally and everything looks fine.

sirluk commented on src/peft/tuners/lora/eva.py:
````diff
@@ -0,0 +1,599 @@
+            padding tokens and should not be used for SVD.
+            `peft.tuners.lora.eva.prepare_model_inputs_fn_language_modeling` is used by default. Any function defined
+            here expects two arguments: `model_input` and `peft_config` and should return a 2d tensor. Default logic:
+            ```
````
fair point, I removed those code snippets from the docstrings

sirluk commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,599 @@
+    return {k: v for k, v in value_dict.items() if len(v) > 1}
+
+
+def recursive_apply(module: torch.nn.Module, device: Union[str, torch.device]):
```
removed

sirluk commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,599 @@
+            f"the following layers were initialized with init_lora_weights=True because they were not found in the eva state_dict: {missing_eva_inits} "
+            f"currently the following lora modules are not supported: {UNSUPPORTED_LORA_MODULES}"
+        )
+    return model
```
That is true, I figured I could do the same as `get_peft_model` as it also modifies a model in place but also returns it. I removed the return statement now

BenjaminBossan reviewed (changes requested):
Thanks for your continued work. The adjustments for saving and loading look good. Also, nice catch with the rank 0 issue.

I have a couple of smaller comments, but the big picture looks good.

sirluk commented on src/peft/tuners/lora/eva.py:
```diff
@@ -571,29 +591,4 @@ def prepare_layer_inputs_fn_language_modeling(layer_input, model_input, layer_na
-            f"currently the following lora modules are not supported: {UNSUPPORTED_LORA_MODULES}"
-        )
-    return model
+    load_eva_state_dict(model, eva_state_dict, adapter_name, _check_valid=False)
```
I did it this way because when call `init_lora_eva_weights` we would otherwise check validity twice, once inside `init_lora_eva_weights` and then again inside `load_eva_state_dict` which is called inside `init_lora_eva_weights`. For that reason I defined the private argument but I am also not super keen on keeping it this way. I could remove it and just accept that the checks are run twice. Or maybe there is a better way to do it.

sirluk commented on tests/test_initialization.py:
```diff
@@ -1742,6 +1746,28 @@ def test_eva_initialization_consistency(self, model, dataset, peft_config):
+        )
+        peft_model = get_peft_model(deepcopy(model), peft_config)
+        sd = get_eva_state_dict(model, peft_config, dataloader)
+        if has_rank_zero:
+            k = "transformer.h.0.attn.c_attn"
+            sd[k] = sd[k][:0]
```
thats correct, it simulates how a rank zero layer in the state dict would look like, a tensor with size 0 in the first dimension

sirluk commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,605 @@
+            module.lora_A[adapter_name].weight.copy_(w)
+            new_target_modules.append(name)
+        else:
+            module.update_layer(
```
it is needed because if we do not have an eva state dict entry for that layer we would like to use standard lora initialization for that layer.

sirluk replied:
Actually when `get_peft_model` is called all adapters would be initialized with `init_lora_weights="eva"` which would only initialized lora B appropriately with zeros but not lora A
```python
def update_layer(...):
        ...
        elif init_lora_weights == "eva":
            nn.init.zeros_(self.lora_B[adapter_name].weight)

BenjaminBossan replied:
Ah yes, I forget about that part, now it makes sense.

sirluk commented on src/peft/tuners/lora/eva.py:
```diff
@@ -487,6 +465,73 @@ def _check_fn(name, module):
+
+    if missing_eva_inits:
+        warnings.warn(
+            f"the following layers were initialized with init_lora_weights=True because they were not found in the eva state_dict: {missing_eva_inits} "
```
done

sirluk commented on tests/test_initialization.py:
```diff
@@ -1742,6 +1746,28 @@ def test_eva_initialization_consistency(self, model, dataset, peft_config):
+    # initialized model can be saved and loaded correctly.
+    @pytest.mark.parametrize("has_rank_zero", [True, False])
+    def test_load_eva_state_dict(self, model, dataset, peft_config, has_rank_zero):
+        tmp_dirname = tempfile.mkdtemp()
```
done

sirluk commented on tests/test_initialization.py:
```diff
@@ -1742,6 +1746,28 @@ def test_eva_initialization_consistency(self, model, dataset, peft_config):
+        load_eva_state_dict(peft_model, sd)
+        # test that models that initialized `load_eva_state_dict` can be saved and loaded correctly
+        peft_model.save_pretrained(tmp_dirname)
+        peft_model = PeftModel.from_pretrained(model, tmp_dirname, torch_device=self.DEVICE, low_cpu_mem_usage=True)
```
done

sirluk commented on tests/test_initialization.py:
```diff
@@ -1742,6 +1746,28 @@ def test_eva_initialization_consistency(self, model, dataset, peft_config):
                 f"is not greater than {self.COSINE_SIMILARITY_THRESHOLD}"
             )
 
+    # Test that the `get_eva_state_dict` function can be used to initialize a model with EVA weights and that the
```
done

sirluk commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,605 @@
+    # update target modules if some lora layers have been removed due to their EVA rank being 0
+    new_target_modules = new_target_modules + missing_eva_inits
+    other_module_names = [n for n, _ in model.named_modules() if n not in new_target_modules]
+    new_target_modules = _find_minimal_target_modules(new_target_modules, other_module_names)
```
done

sirluk commented on src/peft/tuners/lora/config.py:
```diff
@@ -69,6 +69,57 @@ class LoftQConfig:
+            If use_label_mask=True the value to look for to mask out ignored tokens. Default is -100.
+        whiten (`bool`): Apply whitening to singular vectors. Default is False.
+            Whitening has been shown to be beneficial for EVA in the vision domain.
+        use_alpha_pattern (`bool`):
```
done. We also did some checks and it seems in the language modeling setting, adjusting alpha seems to work a bit better. Since I would assume a large share of users use peft for language model tasks I set the default now to `True`

sirluk commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,605 @@
+
+    # when use_alpha_pattern is True, lora scaling factors have been adjusted after the rank redistribution
+    if peft_config.eva_config.use_alpha_pattern:
+        model.peft_config[adapter_name].alpha_pattern = alpha_pattern
```
I would say it is safe now to override as we are incorporating alpha values from `alpha_pattern` into the new `alpha_pattern` dict.

sirluk commented on tests/test_initialization.py:
```diff
@@ -1552,6 +1562,272 @@ def new_state_dict():
+            assert mean_cosine_similarity > self.COSINE_SIMILARITY_THRESHOLD, (
+                f"Mean absolute cosine similarity {mean_cosine_similarity:.4f} "
+                f"is not greater than {self.COSINE_SIMILARITY_THRESHOLD}"
+            )
```
added `test_eva_state_dict_adjust_scaling_factors`

sirluk commented:
@BenjaminBossan thanks for your thoroughness, there were still some edgecases I didnt consider.

Besides the things you pointed out I updated some things additionally (sorry that there are still so many changes in the PR, but I keep discovering potential issues for users or ways to make the api simpler for users)
- Fixed an issue in `get_eva_state_dict` if `LoraConfig` contains a rank pattern. For these cases `max_components` can be different for different layers so this variable is now a dict with keys being the adapter module names.
- removed function `load_eva_state_dict` and integrated functionality into `initialize_lora_eva_weights`. This function now accepts an optional `eva_state_dict` argument.


One question: When loading the eva_state_dict I need to get the module name in the base model based off the module name in the peft model. I this a valid way to do that:
```python
name_in_base_model = name.replace("base_model.model.", "")
```

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,673 @@
+    """
+    Move the inputs to the specified device. Adapted from hf.Trainer.
+    """
+    if isinstance(inputs, Mapping):
```
I think we can add a small optimization here: If we first check if the input has a `.to` method, we can call that (and remove the tensor check below). This is because some HF inputs, like those coming from tokenizers, are a mapping but they also have a `.to` method, which we can make use of.

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,673 @@
+    elif isinstance(inputs, torch.Tensor):
+        return inputs.to(device)
+    else:
+        raise ValueError(f"unsupported input type: {type(inputs)}")
```
I wonder if we really need to raise here. Perhaps the  model somehow takes care of moving to the correct device. I think a warning is sufficient here.

sirluk replied:
sounds good

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,673 @@
+    """
+    devices = list({p.device for p in model.parameters() if p.device.type != "meta"})
+    if len(devices) > 1:
+        raise ValueError("model has multiple devices")
```
I wonder if we should really raise an error here. Maybe the user has some kind of custom model that can correctly deal with the devices, and raising here would only stop them unnecessarily. How about giving a warning and skipping the `move_inputs_to_device(inputs, device)` if `device` could not be determined?

sirluk replied:
sounds good

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,673 @@
+        "use_rslora": peft_config.use_rslora,
+        "use_dora": peft_config.use_dora,
+    }
+    # rank_pattern_keys() = list(chain(peft_config.rank_pattern.keys(), peft_config.alpha_pattern.keys()))
```
Can this be removed?

BenjaminBossan commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,673 @@
+    return {k: v for k, v in value_dict.items() if len(v) > 1}
+
+
+def get_target_name_key(name, pattern_keys):
```
This can be replaced by `get_pattern_key` if #2195 is merged first, right? Let's wait with this PR for the other one then.

sirluk replied:
yes correct!

BenjaminBossan reviewed (changes requested):
Thanks for your continued work on this PR, this is really excellent work.

Regarding your latest changes, they make sense to me, thanks for explaining. I did another check and flagged some smaller issues, mainly around making this approach more flexible for custom models. Otherwise, this PR is good to be merged.

> One question: When loading the eva_state_dict I need to get the module name in the base model based off the module name in the peft model. I this a valid way to do that:

Yes, I think the way it's implemented is good. For LoRA, we will always have this prefix.

sirluk commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,673 @@
+from .layer import Embedding, LoraLayer, _ConvNd
+
+
+UNSUPPORTED_LORA_MODULES = (Embedding, _ConvNd)
```
fair point, the only advantage is having a better error message, I can invert this to `SUPPORTED_LORA_MODULES`

sirluk replied:
just was reminded during testing that the actual reason why I did it this way is to automatically include bnb linear layers. Otherwise I would need to do a conditional import I guess if bnb is available. Would you still say its better to do it this way?

BenjaminBossan replied:
No, that's why I mentioned

> Eva would generally work with newly added quantization methods

For me, this was more to verify the reasoning here When new LoRA layers are added that are not supported here, we'll most likely forget to update this constant, but it's not the end of the world. However, I think we can make it a bit more likely no to miss this if we move `UNSUPPORTED_LORA_MODULES` to `utils/constants.py` and rename it to `UNSUPPORTED_EVA_LORA_MODULES` or something like this.

sirluk replied:
moving this to `constants.py` doesn't work unfortunately due to some circular import issue with `PeftConfig`

BenjaminBossan replied:
Hmm, any idea how that can be? In the [line above](https://github.com/huggingface/peft/pull/2142/files#diff-36e504fe6dcda50b98764f8961b977a65a976f187bc391385299fba1fc07704bR29), we already import from `peft.utils.constants`, so it's unclear to me how a circular import code arise here.

sirluk replied:
It has something to do with me now needing to import from lora/layer.py. I think the circular dependency goes something like this
- in constants.py we now import from lora/layer.py
- lora/layer.py imports from lora/config.py
- lora/config.py imports from peft/config.py
- peft/config.py imports from utils/other.py
- utils/other.py imports from constants.py

BenjaminBossan replied:
Ah yes, that makes sense. Then I wonder if we can define the constant inside of `lora/__init__.py`. Just somewhere more visible than here :)

sirluk replied:
hmm unfortunately this does not fix the issue. Putting the variable in `lora/__init__.py` still raises an error.

BenjaminBossan replied:
All right, then let's just keep it as is.

sirluk commented on src/peft/tuners/lora/eva.py:
```diff
@@ -0,0 +1,673 @@
+        raise ValueError("`initialize_lora_eva_weights` can only be used with `init_lora_weights='eva'`")
+
+    # compute svd
+    if eva_state_dict is None:
```
But for `init_lora_eva_weights` the dataloader can actually be `None` if state dict is provided. I added a `None` check to `_get_eva_state_dict` where the dataloader is actually needed

BenjaminBossan replied:
> But for init_lora_eva_weights the dataloader can actually be None if state dict is provided.

Well, here we already checked that `eva_state_dict is None`, so that would be covered.

> I added a `None` check to `_get_eva_state_dict` where the dataloader is actually needed

That's also fine, then the type annotation should be updated to `dataloader: Optional[Iterable]`

sirluk replied:
sorry, which type annotation do you mean, I saw its not updated in the docstring of `init_lora_eva_weights` and added it there.

In `get_eva_state_dict` and `_get_eva_state_dict` its not really optional as it will always raise an error.

BenjaminBossan replied:
IMO it is optional because the function accepts and expects that `None` could be passed. I would expect a type checker to complain otherwise, as the `get_eva_state_dict` call here passes `dataloader`, which according to this function could be `None`.

It's no big deal overall, I personally don't use type checkers on this code base, but I know that some users complain sometimes because their IDE warns about the wrong type being used.

sirluk replied:
Okay that makes sense, I moved to `None` check to `init_lora_eva_weights` which was your original suggestion anyway. So in the other functions the `dataloader` argument would not be optional anymore now.

sirluk commented:
@BenjaminBossan, thanks integrated you suggestions and left some comments to the open threads:
- when calling `update_layer` is needed
- unsupported or supported lora modules
- `None` check for dataloader

BenjaminBossan reviewed (commented):
Thanks for the updates, I think we're almost good to go. Just 2 small comments from my side.

> when calling update_layer is needed

Sorry, I can't find the comment, could you please paste it again? Just to summarize, my main concern is that we may call `module.update_layer` even if it's not necessary, namely when the module was _not_ initialized with `low_cpu_mem_usage=True` (i.e. when it is an real device, not meta, we would not need to initialize it).

sirluk commented:
@BenjaminBossan this is the [dicussion](https://github.com/huggingface/peft/pull/2142#discussion_r1827647640) about `update_layer`.

For the other two open conversation I added a comment directly to the conversation.

BenjaminBossan reviewed (commented):
I replied to your latest comments. Also, there is a small merge conflict, could you please take care? Thanks.

sirluk commented:
Great, thanks for your feedback, I made some small changes, I think the only open issue is the potential circular import `constants.py`.

Small note: I changed the default value of `rho` back to `1.0` as this works well in most settings.

BenjaminBossan commented:
@sirluk Some Eva tests are failing:

https://github.com/huggingface/peft/actions/runs/11698806717/job/32754904346?pr=2142#step:6:1195

Is it possible that you tested this with a GPU and there it's fast enough, but on the CPU CI runners, it's too slow? Maybe we should remove the timing assertions, as they tend to be flaky?

sirluk commented:
Thanks for pointing this out, I was indeed testing on a gpu machine so it was automatically running everything on gpu. I removed the speed test now.

BenjaminBossan reviewed (approved):
Thanks a lot for ironing out the last kinks. The PR now LGTM. Really excellent work, thanks for that and for your patience.

I hope that many users will give it a try, as it looks quite promising.

sirluk commented:
Great, thanks for your great feedback throughout the process!

julien-c commented:
Great PR!

BenjaminBossan commented:
@sirluk I just wanted to make you aware of PR #2231 which introduces a (at least superficially) similar method to EVA, which also benefited from our discussion here to improve the API. The PR includes a [small table](https://github.com/huggingface/peft/blob/9d4a5200dd340fecd89b9be4178825095b18c508/examples/corda_finetuning/README.md#comparison-with-alternative-methods) that summarizes a couple of methods including EVA, maybe you can double check that it's accurate.

sirluk commented:
@BenjaminBossan, thanks for making us aware of this recent PR. It is indeed very relevant and interesting to us as well. I'm glad our discussion benefits the integration of other data-driven methods into PEFT. We believe the table is not entirely correct with respect to EVA. From their paper, "Knowledge Maintenance mode" in CorDA simply means using the smallest components for initialization, rather than the largest ones. This is necessary for methods like CorDA and PiSSA because they modify the base model weights, meaning the base model cannot be restored after fine-tuning. We intentionally followed the design of LoRA and left the base model untouched as this allows restoring the base model's weights and therefore full knowledge preservation. We just recently  added a discussion paragraph on that in the updated version of our paper which will be on arXiv soon.
(tagging @paischer101 as he is the other joint first-author of EVA)

BenjaminBossan commented:
@sirluk Thanks for the reply. Somehow I didn't get a notification for your message or missed it, which is why I didn't respond earlier.

The CorDA PR is almost finished but there is still an opportunity to ask the authors to correct the table. From your comment, it was not quite clear to me what exactly was not entirely correct. Could you perhaps post directly on that PR and mention what could be improved?

paischer101 commented:
@BenjaminBossan thanks for the response. Since EVA does not alter the base model weights, it is able to recover the base model and therefore allows full knowledge maintenance, i.e. the last column in the table should be a "yes" for EVA as well. I will make a post on the PR to clarify.

sirluk commented:
Hey @BenjaminBossan , just wanted to let you know there is a mistake in the latest peft release notes describing EVA. It says that we use "SVD of the base layer weights" which is not correct. EVA is a data driven method that performs incremental SVD on minibatches of finetuning data. The result is used to initilize LoRA and perform an adaptive rank redistribution. Could you maybe change that so that potential users are not confused?

githubnemo commented:
@sirluk Thanks for notifying us. I changed the release text accordingly.
